distro/fedora: Refactor based on rhel-90

jordigilh opened the pull request:
This pull request includes:

- [X] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
- [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

This PR contains the code for Fedora 34,35 and 36 that uses the v2 API. The process to migrate the code has been taking RHEL 9.0 as the baseline and then migrating the packages from v1 (fedora33). With the changes, new image types are now available for Fedora:
* fedora-iot-commit for aarch64
* fedora-iot-container for x86_64 and aarch64
* fedora-iot-installer for x86_64 and aarch64

It also removes the old v1 code ([distro/fedora33](https://github.com/osbuild/osbuild-composer/tree/main/internal/distro/fedora33)) and splits each fedora version into its own directory:
* distro/fedora34
* distro/fedora35
* distro/fedora36

Notes:
* Extended function signature for `NewKickstartStageOptions` to include the osName as an argument. 
https://github.com/osbuild/osbuild-composer/blob/c2c1f39d78ed943913e8ee2bb77d4aa835e6e0fa/internal/osbuild2/kickstart_stage.go#L41-L47
This change allows to reuse the function for both `rhel` and `fedora` use cases, since it was previously hardcoded to `rhel`:
  https://github.com/osbuild/osbuild-composer/blob/c2c1f39d78ed943913e8ee2bb77d4aa835e6e0fa/internal/osbuild2/kickstart_stage.go#L63-L68
*  In order to make it work I had to make some tweaks:
  Unpack the `cloud server tools` package group and list instead the containing packages, removing the `fedora-release-cloud` from the default packages. This was required because it collided with another `fedora-release-identity` coming from the blueprint. The identity used is [fedora-release-identity-basic](https://github.com/jordigilh/osbuild-composer/blob/58beac0cbfc6d893eba1b269ff63bf7a880f5655/test/data/manifests/fedora_35-x86_64-qcow2-boot.json#L2058).  

Additionally, when running in `x86_64`, `syslinux-extlinux` is added  to the `cloud server tools` package set, as it is listed as part of the package, but not for aarch64:

x86_64:
```bash
[root@helios08 ~]# uname -i
x86_64
root@helios08 ~]# dnf group info "Cloud Server Tools"
Last metadata expiration check: 0:03:59 ago on Fri 18 Mar 2022 09:42:43 PM IST.
Group: Cloud Server Tools
 Description: Software for servers and system images that run in "cloud" environments
 Mandatory Packages:
   cloud-init
   cloud-utils-growpart
   dracut-config-generic
   grubby
   rsync
   syslinux-extlinux
   tar
``` 

aarch64:
```bash
[root@ampere-hr330a-01 ~]# uname -i
aarch64
[root@ampere-hr330a-01 ~]# dnf group info "cloud server tools"
Last metadata expiration check: 0:05:05 ago on Fri 18 Mar 2022 03:42:33 PM EDT.
Group: Cloud Server Tools
 Description: Software for servers and system images that run in "cloud" environments
 Mandatory Packages:
   cloud-init
   cloud-utils-growpart
   dracut-config-generic
   grubby
   rsync
   tar
```

* `Core` package is listed as group dependency. I tried to unpack it and use the listed packages but then I found package dependency issues for x86_64 and aarch64, which doesn't happen when I use  the group:

For x86_64:
```bash
panic: Could not depsolve: DNF error occured: MarkingErrors: Error occurred when marking packages for installation: Problems in request:
missing packages: dracut-config-rescue, zram-generator-defaults
```

For aarch64:
```bash
panic: Could not depsolve: DNF error occured: MarkingErrors: Error occurred when marking packages for installation: Problems in request:
missing packages: dracut-config-rescue, firewalld, plymouth, zram-generator-defaults
```

* I kept the file system as `ext4`, as well as the disk size for the image types defined in v1, but kept the partition structure from v2 since it is not a straight forward migration. It felt better to keep the v2 partition structure as it was already proven to work.  Let me know if you'd prefer to keep the partitions as close as v1 and I'll bring the changes the best I can.

@thozza @achilleas-k @teg please take a look at your discretion.

henrywang commented:
Tests for `fedora-iot-container` and `fedora-iot-installer` are ready to push here, it depends on https://github.com/osbuild/gitlab-ci-terraform/pull/60 to add `large` fedora 35 runner.

thozza commented:
Thanks a lot @jordigilh .

I still need to do a full review, but at the first glance, I'm not sure that having a separate distro definition for each Fedora version is what we would like to have in composer. At least for now. I would prefer composer to have only a single Fedora distro definition. What do you think @achilleas-k, @gicmo, @teg ?

Are there currently any differences between the F34, F35 and F36 definitions?

henrywang commented:
podman bug [bz#2065292](https://bugzilla.redhat.com/show_bug.cgi?id=2065292) blocked `Container` test, and the rest of tests can't be run in this case.

gicmo commented:
Indeed, thanks a ton @jordigilh !!

> I still need to do a full review, but at the first glance, I'm not sure that having a separate distro definition for each Fedora version is what we would like to have in composer. At least for now. I would prefer composer to have only a single Fedora distro definition. What do you think @achilleas-k, @gicmo, @teg ?

Yeah, I think a single pipeline would be the aim/goal. It might not always work, e.g. if something changes fundamentally and we don't want to have an explosion of conditionals for old craft, it might makes sense to duplicate the code, esp. if this would correspond with a RHEL branch.

jordigilh commented:
> Thanks a lot @jordigilh .
> 
> I still need to do a full review, but at the first glance, I'm not sure that having a separate distro definition for each Fedora version is what we would like to have in composer. At least for now. I would prefer composer to have only a single Fedora distro definition. What do you think @achilleas-k, @gicmo, @teg ?
> 
> Are there currently any differences between the F34, F35 and F36 definitions?

Only for F34 and that's a package group that would need to be removed, compared to F35 and F36. I split because I considered it easier to remove the directory when the version EOL'd, rather than have to dig into the code and remove the parts belonging to that version.

I'll go ahead and merge them into one directory, following the existing code.

achilleas-k commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,913 @@
+		filename:    "installer.iso",
+		mimeType:    "application/x-iso9660-image",
+		packageSets: map[string]packageSetFunc{
+			// TODO: non-arch-specific package set handling for installers
+			// This image type requires build packages for installers and
+			// ostree/edge.  For now we only have x86-64 installer build
+			// package sets defined.  When we add installer build package sets
+			// for other architectures, this will need to be moved to the
+			// architecture and the merging will happen in the PackageSets()
+			// method like the other sets.
```
This TODO is obsolete.  I realise it was copied from the `rhel90` package, but it's obsolete there as well :smile: 
(I should clean that up at some point)

jordigilh replied:
I'll remove it. Thanks!

jordigilh replied:
Done, removed

achilleas-k commented:
Looks like there are a few F35 manifests that are failing in CI (extra/missing packages)

jordigilh commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,913 @@
+}
+
+// distribution objects without the arches > image types
+var distroMap = map[string]distribution{
+	fedora34Distribution: {
+		name:               fedora34Distribution,
+		product:            "Fedora",
+		osVersion:          "34",
+		releaseVersion:     "34",
+		modulePlatformID:   "platform:fedora",
+		vendor:             "fedora",
+		ostreeRefTmpl:      "fedora/34/%s/edge",
+		isolabelTmpl:       "Fedora-34-BaseOS-%s",
+		runner:             "org.osbuild.fedora33",
+		defaultImageConfig: defaultDistroImageConfig,
+	},
+	fedora35Distribution: {
+		name:               fedora35Distribution,
+		product:            "Fedora",
+		osVersion:          "35",
+		releaseVersion:     "35",
+		modulePlatformID:   "platform:fedora",
+		vendor:             "fedora",
+		ostreeRefTmpl:      "fedora/35/%s/edge",
+		isolabelTmpl:       "Fedora-35-BaseOS-%s",
+		runner:             "org.osbuild.fedora33",
+		defaultImageConfig: defaultDistroImageConfig,
+	},
+	fedora36Distribution: {
+		name:               fedora36Distribution,
+		product:            "Fedora",
+		osVersion:          "36",
+		releaseVersion:     "36",
+		modulePlatformID:   "platform:fedora",
+		vendor:             "fedora",
+		ostreeRefTmpl:      "fedora/36/%s/edge",
+		isolabelTmpl:       "Fedora-36-BaseOS-%s",
+		runner:             "org.osbuild.fedora33",
+		defaultImageConfig: defaultDistroImageConfig,
+	},
```
Makes sense. Thanks :smile:

jordigilh replied:
Done, updated each distribution to use its own runner:
* F34: `org.osbuild.fedora34`
* F35: `org.osbuild.fedora35`
* F36: `org.osbuild.fedora36`

I've also changed the platformID: it was `platform:fedora` for all 3, but in the failed test I noticed that it was expecting (?) to be `platform:f<number>`.
* F34: `platform:f34`
* F35: `platform:f35`
* F36: `platform:f36`

I'll keep an eye on the tests when they eventually run (it takes a very very long time for those to get triggered).

thozza reviewed (commented):
Thanks a LOT for your work on this. ❤️   I know that it must have been a lot to juggle the move to V2 and trying to keep Fedora definitions close to what they were when using V1. 

In general, this PR looks great and there are mostly cosmetic issues. I added "a few" comments... Sorry 🙄  Here are some more general notes:

- The Distribution, Arch and ImageType are defined as interfaces. We are not really good at using interface methods, instead of unexported structure members, so copying any such code from `rhel-90` is definitely acceptable. However, for the newly added code, please try to use the interfaces if possible. I commented a few places like this, but feel free to ignore comments to parts which are exactly the same in `rhel-90` code. https://github.com/osbuild/osbuild-composer/blob/f1006c5adb9d8fd6cd09feaf739bfb6d0df7d35e/internal/distro/distro.go#L37-L127
- There are some discrepancies in the image configurations, which are probably just missed changes. e.g. the default timezone, default image size of `qcow2`, etc. I assume that you have been reviewing the differences in the image tests which you regenerated (the huge JSON files in `test/data/manifests`). I would like to point out, that although the manifests changed completely, the `"image-info"` part of the file is agnostic to the manifest format and is a reasonably good "source of truth" for the image configuration. So reviewing at least changes to that section for previously supported images is very valuable. After the review, I think that you actually reviewed it, at least partially.
- It is a good idea to keep most of the functions copied from `rhel-90` definition as identical to `rhel-90` as possible (exceptions apply). The reason is that we are continuously de-duplicating copies of various functions with the aim to eventually share most of the code across various distro definitions. Keeping the code as similar as possible will make it easier to de-duplicate it eventually.

jordigilh commented on test/data/manifests/fedora_34-aarch64-oci-boot.json:
```diff
@@ -10042,7 +10485,7 @@
       }
     },
-    "timezone": "UTC",
+    "timezone": "New_York",
```
I'll change it back to UTC. Thanks for pointing it out.

jordigilh commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,906 @@
+	// the layout is converted to LVM so we need to corresponding packages
+	if !t.rpmOstree {
+
+		pt, exists := t.basePartitionTables[t.arch.name]
+		if !exists {
+			panic(fmt.Sprintf("unknown architecture with boot type: %s %s", t.arch.Name(), t.bootType))
+		}
```
A safeguard in case the partition for the given architecture does not exist. If you think this case would never happen, I'm happy to ignore the check :smile: .

thozza replied:
I should have been more precise... 😄 I meant not using the interface function. The check and `panic()` is definitely a good addition 😉

jordigilh replied:
ah yeah sorry, my bad :smile: . Interfaces are the way to go. I see room for improvement on the `t.Arch()....`, though the changes would impact code across distributions and I'd rather keep this PR constrained to Fedora only. I'll raise a new PR to extend the interface to cover other cases in all existing distributions :smile:

jordigilh commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,906 @@
+	options distro.ImageOptions,
+	rng *rand.Rand,
+) (*disk.PartitionTable, error) {
+	basePartitionTable, exists := t.basePartitionTables[t.arch.name]
+	if !exists {
+		panic(fmt.Sprintf("unknown architecture with boot type: %s %s", t.arch.Name(), t.bootType))
+	}
```
none. I'll return the error. Thanks for pointing it out.

jordigilh commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,906 @@
+}
+
+func (t *imageType) PartitionType() string {
+	basePartitionTable, exists := t.basePartitionTables[t.arch.name]
```
agreed. Let me change that.

jordigilh commented on internal/distro/fedora/pipelines.go:
```diff
@@ -0,0 +1,670 @@
+	platform := arch.legacy
+	if platform != "" {
+		return osbuild.NewGrub2InstStage(osbuild.NewGrub2InstStageOption(filename, pt, platform))
+	}
```
understood. I'll add the code back.

jordigilh commented on internal/distro/fedora/package_sets.go:
```diff
@@ -0,0 +1,758 @@
+
+// OS package sets
+
+func coreOsCommonPackageSet(t *imageType) rpmmd.PackageSet {
+	ps := rpmmd.PackageSet{
+		Include: []string{
+			"@core",
+		},
+	}
+
+	return ps
+}
```
understood. Will update as suggested.

jordigilh commented on internal/distro/fedora/package_sets.go:
```diff
@@ -0,0 +1,758 @@
+
+	}
+
+	r, err := strconv.Atoi(t.arch.distro.releaseVersion)
```
ack

jordigilh commented on internal/distro/fedora/distro_test.go:
```diff
@@ -0,0 +1,731 @@
+	"github.com/osbuild/osbuild-composer/internal/distro/fedora"
+)
+
+type rhelFamilyDistro struct {
```
good catch!.

jordigilh commented on internal/distro/fedora/distro_test.go:
```diff
@@ -0,0 +1,731 @@
+	distro distro.Distro
+}
+
+var rhelFamilyDistros = []rhelFamilyDistro{
```
yep

jordigilh commented on test/data/manifests/fedora_34-aarch64-ami-boot.json:
```diff
@@ -9962,27 +10342,37 @@
-        "label": "EFI-SYSTEM",
-        "partuuid": "02C1E068-1D2F-4DA3-91FD-8DD76A955C9D",
-        "size": 498073600,
+        "label": null,
```
I'll take a look. Thanks for pointing it out :)

jordigilh replied:
it seems like the value is fetched from the `sfidsk` call and in this case it does not return a label.
```json
{
   "partitiontable": {
      "label":"gpt",
      "id":"D209C89E-EA5E-4FBD-B161-B461CCE297E0",
      "device":"beaf03ae-1586-4308-97f0-cda78e72f160-image.raw",
      "unit":"sectors",
      "firstlba":2048,
      "lastlba":12582878,
      "sectorsize":512,
      "partitions": [
         {"node":"beaf03ae-1586-4308-97f0-cda78e72f160-image.raw1", "start":2048, "size":2048, "type":"21686148-6449-6E6F-744E-656564454649", "uuid":"FAC7F1FB-3E8D-4137-A512-961DE09A5549"},
         {"node":"beaf03ae-1586-4308-97f0-cda78e72f160-image.raw2", "start":4096, "size":409600, "type":"C12A7328-F81F-11D2-BA4B-00A0C93EC93B", "uuid":"68B2905B-DF3E-4FB3-80FA-49D1E773AA33"},
         {"node":"beaf03ae-1586-4308-97f0-cda78e72f160-image.raw3", "start":413696, "size":1024000, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"CB07C243-BC44-4717-853E-28852021225B"},
         {"node":"beaf03ae-1586-4308-97f0-cda78e72f160-image.raw4", "start":1437696, "size":11145183, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"6264D520-3FB9-423F-8AB8-7A0A8E3D3562"}
      ]
   }
}
```

Here is the code:
https://github.com/osbuild/osbuild-composer/blob/72019740c24351aa7719f05d81e66f849e30ce81/tools/image-info#L247-L249

jordigilh commented:
> Thanks a LOT for your work on this. heart I know that it must have been a lot to juggle the move to V2 and trying to keep Fedora definitions close to what they were when using V1.
> 
> In general, this PR looks great and there are mostly cosmetic issues. I added "a few" comments... Sorry roll_eyes Here are some more general notes:
> 
>     * The Distribution, Arch and ImageType are defined as interfaces. We are not really good at using interface methods, instead of unexported structure members, so copying any such code from `rhel-90` is definitely acceptable. However, for the newly added code, please try to use the interfaces if possible. I commented a few places like this, but feel free to ignore comments to parts which are exactly the same in `rhel-90` code. https://github.com/osbuild/osbuild-composer/blob/f1006c5adb9d8fd6cd09feaf739bfb6d0df7d35e/internal/distro/distro.go#L37-L127
> 
>     * There are some discrepancies in the image configurations, which are probably just missed changes. e.g. the default timezone, default image size of `qcow2`, etc. I assume that you have been reviewing the differences in the image tests which you regenerated (the huge JSON files in `test/data/manifests`). I would like to point out, that although the manifests changed completely, the `"image-info"` part of the file is agnostic to the manifest format and is a reasonably good "source of truth" for the image configuration. So reviewing at least changes to that section for previously supported images is very valuable. After the review, I think that you actually reviewed it, at least partially.
> 
>     * It is a good idea to keep most of the functions copied from `rhel-90` definition as identical to `rhel-90` as possible (exceptions apply). The reason is that we are continuously de-duplicating copies of various functions with the aim to eventually share most of the code across various distro definitions. Keeping the code as similar as possible will make it easier to de-duplicate it eventually.

Thanks for the lengthily review and feedback! I have been looking at the packages in the manifests mostly, the only thing I noticed so far is the change to the `fedora-release-identity` package due to conflict that I mentioned already elsewhere. With the change in the structure it is hard to review them, and even harder without knowing exactly what is supposed to be kept and what should be changed with the migration. So thanks again for pointing out about the `image-info` section, I will take another look and see what I need to change to match the old manifest :smile:

teg commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,906 @@
+	}
+
+	// Image Definitions
+	iotCommitImgType := imageType{
+		name:        "fedora-iot-commit",
+		nameAliases: []string{"iot-commit"},
```
Would love to unify, and at least not include fedora/rhel in the name. But I think that requires more discussion with the Fedora IoT SIG, so totally fine to leave that for the future.

jordigilh replied:
With no objections so far, I'll leave the alias for `iot-{commit,container,installer}`, but I agree that having a single image type across distributions makes sense at the end.

Resolving the comment.

jordigilh commented:
> > Thanks a LOT for your work on this. heart I know that it must have been a lot to juggle the move to V2 and trying to keep Fedora definitions close to what they were when using V1.
> > In general, this PR looks great and there are mostly cosmetic issues. I added "a few" comments... Sorry roll_eyes Here are some more general notes:
> > ```
> > * The Distribution, Arch and ImageType are defined as interfaces. We are not really good at using interface methods, instead of unexported structure members, so copying any such code from `rhel-90` is definitely acceptable. However, for the newly added code, please try to use the interfaces if possible. I commented a few places like this, but feel free to ignore comments to parts which are exactly the same in `rhel-90` code. https://github.com/osbuild/osbuild-composer/blob/f1006c5adb9d8fd6cd09feaf739bfb6d0df7d35e/internal/distro/distro.go#L37-L127
> > 
> > * There are some discrepancies in the image configurations, which are probably just missed changes. e.g. the default timezone, default image size of `qcow2`, etc. I assume that you have been reviewing the differences in the image tests which you regenerated (the huge JSON files in `test/data/manifests`). I would like to point out, that although the manifests changed completely, the `"image-info"` part of the file is agnostic to the manifest format and is a reasonably good "source of truth" for the image configuration. So reviewing at least changes to that section for previously supported images is very valuable. After the review, I think that you actually reviewed it, at least partially.
> > 
> > * It is a good idea to keep most of the functions copied from `rhel-90` definition as identical to `rhel-90` as possible (exceptions apply). The reason is that we are continuously de-duplicating copies of various functions with the aim to eventually share most of the code across various distro definitions. Keeping the code as similar as possible will make it easier to de-duplicate it eventually.
> > ```
> 
> Thanks for the lengthily review and feedback! I have been looking at the packages in the manifests mostly, the only thing I noticed so far is the change to the `fedora-release-identity` package due to conflict that I mentioned already elsewhere. With the change in the structure it is hard to review them, and even harder without knowing exactly what is supposed to be kept and what should be changed with the migration. So thanks again for pointing out about the `image-info` section, I will take another look and see what I need to change to match the old manifest smile

Update:
Combed the manifests and found a few discrepancies. I was able to clean most of them but a few eluded me:
*  The rpm`glibc-minimal-langpack` is now added to the dependencies, I suspect it has to do with the kernel, but I am not sure. I was not able to remove it.
* `fedora-release-identity-*` is still not aligned with the proper package due to conflict. I'm a bit worried that this will cause other issues down the road as now some of the fields in the `os-release` struct will not match. Example for `qcow2`:
Expected:
```json
      "PRETTY_NAME": "Fedora 34 (Cloud Edition)",
      "VARIANT": "Cloud Edition",
      "VARIANT_ID": "cloud",
      "VERSION": "34 (Cloud Edition)",
```
Actual:
```json
      "PRETTY_NAME": "Fedora 34 (Thirty Four)",
      "VERSION": "34 (Thirty Four)",
```

thozza commented:
Thanks a lot for the update!

> Update: Combed the manifests and found a few discrepancies. I was able to clean most of them but a few eluded me:
> 
> * The rpm`glibc-minimal-langpack` is now added to the dependencies, I suspect it has to do with the kernel, but I am not sure. I was not able to remove it.

This is not really a big issue IMO.

> * `fedora-release-identity-*` is still not aligned with the proper package due to conflict. I'm a bit worried that this will cause other issues down the road as now some of the fields in the `os-release` struct will not match. Example for `qcow2`:
>   Expected:
> 
> ```json
>       "PRETTY_NAME": "Fedora 34 (Cloud Edition)",
>       "VARIANT": "Cloud Edition",
>       "VARIANT_ID": "cloud",
>       "VERSION": "34 (Cloud Edition)",
> ```
> 
> Actual:
> 
> ```json
>       "PRETTY_NAME": "Fedora 34 (Thirty Four)",
>       "VERSION": "34 (Thirty Four)",
> ```

This is IMO a real issue. Some packages may be applying configuration during the installation conditionally based on the variant on which they are installed. @achilleas-k is currently working on refactoring of the `dnf-json` tool, which we use for dependency solving a DNF transaction. We also have couple of ideas how to reword the dependency solving to prevent the type of issues that you've encountered.

There are two options from my point of view:

1. Blocking this PR until we rework the depsolving in composer (can realistically take 1-2 weeks to be fully merged).
2. Get this version of Fedora definition in and change it once we rework depsolving. It would be good to add commented out lines of code which should be used once we fix depsolving, as part of this PR.

I'm OK with both approaches, since I know that you've already spent quite a lot of time on this PR and the issue would be present in composer for a limited time and won't affect our Fedora plans much. @gicmo any preference ^

jordigilh commented on tools/test-case-generators/format-request-map.json:
```diff
@@ -494,11 +525,7 @@
           }
         ],
         "modules": [],
-        "groups": [
```
@thozza @achilleas-k I had to empty this group for this target as it added extra packages that were not part of the v1 for Fedora. It turns out actually, that the qcow2 already incorporates the `@core` package as part of it's package list, so no harm done here. Let me know if you want me to put it back and I'll do so :smile: 

This also raises the question, I wonder, that adding the same group package through the blueprint somehow incorporates extra rpms.

achilleas-k replied:
Yes, this is a known issue with the way we depsolve the two sets of packages (base and blueprint).  We're working on fixing it but it's a pretty big change.
In short, the blueprint packages and depsolved separately without restrictions (no excluded packages), whereas in the base package set we exclude some weak dependencies.  So if `@core` is added to the blueprint, it pulls in all those optional packages as well.
See https://github.com/osbuild/osbuild-composer/issues/2386

jordigilh replied:
got it. I'll remove it for now to match the rpms from old fedora manifests, but please let me know if you'd like it back.

jordigilh commented:
> Thanks a lot for the update!
> 
> > Update: Combed the manifests and found a few discrepancies. I was able to clean most of them but a few eluded me:
> > 
> > * The rpm`glibc-minimal-langpack` is now added to the dependencies, I suspect it has to do with the kernel, but I am not sure. I was not able to remove it.
> 
> This is not really a big issue IMO.
> 
> > * `fedora-release-identity-*` is still not aligned with the proper package due to conflict. I'm a bit worried that this will cause other issues down the road as now some of the fields in the `os-release` struct will not match. Example for `qcow2`:
> >   Expected:
> > 
> > ```json
> >       "PRETTY_NAME": "Fedora 34 (Cloud Edition)",
> >       "VARIANT": "Cloud Edition",
> >       "VARIANT_ID": "cloud",
> >       "VERSION": "34 (Cloud Edition)",
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Actual:
> > ```json
> >       "PRETTY_NAME": "Fedora 34 (Thirty Four)",
> >       "VERSION": "34 (Thirty Four)",
> > ```
> 
> This is IMO a real issue. Some packages may be applying configuration during the installation conditionally based on the variant on which they are installed. @achilleas-k is currently working on refactoring of the `dnf-json` tool, which we use for dependency solving a DNF transaction. We also have couple of ideas how to reword the dependency solving to prevent the type of issues that you've encountered.
> 
> There are two options from my point of view:
> 
>     1. Blocking this PR until we rework the depsolving in composer (can realistically take 1-2 weeks to be fully merged).
> 
>     2. Get this version of Fedora definition in and change it once we rework depsolving. It would be good to add commented out lines of code which should be used once we fix depsolving, as part of this PR.
> 
> 
> I'm OK with both approaches, since I know that you've already spent quite a lot of time on this PR and the issue would be present in composer for a limited time and won't affect our Fedora plans much. @gicmo any preference ^

Agreed. Personally, I'd go for 2 for now since I already had to make some changes in my code due to merge conflicts as work laded in the main branch while I was still working on my PR, but I will accept whatever decision is made :smile: .

gicmo commented:
> Agreed. Personally, I'd go for 2 for now since I already had to make some changes in my code due to merge conflicts as work laded in the main branch while I was still working on my PR, but I will accept whatever decision is made 😄 .

Totally fine with me!

gicmo commented on tools/test-case-generators/distro-arch-imagetype-map.json:
```diff
@@ -302,4 +312,4 @@
       "tar"
     ]
   }
-}
```
😱 whitespace change! ;)

jordigilh replied:
I'll add it back :smiley:

achilleas-k reviewed (changes requested):
Came back to review again now that it seems this is complete and also CI is green and noticed that all the changes are contained in a single commit.  It's a little understandable, for the most part, because it's a complete rewrite and it might be hard to separate it into meaningful partial changes, but some things can definitely be split out and don't belong together.

Specifically, the changes in `internal/blueprint/customizations_test.go` could be a separate commit.  The creation of the `fedora` package could be separate from the deletion of the `fedora33` package and then tests can be changed when the new package is imported?  I'm not sure what the best or nicest split is and if others have other thoughts, but I feel uncomfortable seeing this many changes in a single commit.

achilleas-k commented:
> @jordigilh @achilleas-k Should I add test in this PR or in separate one? Thanks.

Let's test here

achilleas-k commented:
> @achilleas-k I got error `failed to create osbuild manifest: boot ISO image type \"fedora-iot-installer\" does not support blueprint customizations` in `fedora-iot-installer` test. This PR does not include code change in PR #2375? Thanks.

Yes, it seems it doesn't.

henrywang commented:
Thanks @achilleas-k.
@jordigilh If you would like to add this feature in separate PR, I'll remove test commit in this PR. Thanks.

jordigilh commented:
> Thanks @achilleas-k. @jordigilh If you would like to add this feature in separate PR, I'll remove test commit in this PR. Thanks.

No worries, I'll add the feature in this PR. No need to remove your commit.

jordigilh commented:
> > Thanks @achilleas-k. @jordigilh If you would like to add this feature in separate PR, I'll remove test commit in this PR. Thanks.
> 
> No worries, I'll add the feature in this PR. No need to remove your commit.

@henrywang PR updated with support to allow user and group customizations in iot-installer.

henrywang commented:
@jordigilh @achilleas-k @ondrejbudai All GCP tests are failed due to 
```ERROR: (gcloud.compute.instances.create) Could not fetch resource: - Quota 'IN_USE_ADDRESSES' exceeded.  Limit: 8.0 in region us-west1.```

achilleas-k commented on internal/distro/fedora/pipelines.go:
```diff
@@ -171,8 +171,12 @@ func iotInstallerPipelines(t *imageType, customizations *blueprint.Customization
+		return nil, err
+	}
+	ksUsers := len(customizations.GetUsers())+len(customizations.GetGroups()) > 0
+	pipelines = append(pipelines, *anacondaTreePipeline(repos, installerPackages, kernelVer, archName, d.product, d.osVersion, "iot", ksUsers))
```
I downloaded the official Fedora IoT ISO and checked the `.buildstamp` file and the Variant there is capitalised as `IoT`.  Probably not important though?

jordigilh replied:
I will change it. No worries :smile:

achilleas-k commented on internal/osbuild2/kickstart_stage.go:
```diff
@@ -44,7 +44,8 @@ func NewKickstartStageOptions(
 	ostreeURL string,
-	ostreeRef string) (*KickstartStageOptions, error) {
+	ostreeRef string,
+	osName string) (*KickstartStageOptions, error) {
```
Noticed this change in the callers before seeing it here and figuring out why it happened.  If there's a rebase at any point for whatever reason, could you please mention it in the commit message?

jordigilh replied:
Acknowledged. I have updated the description to explain the change in the function signature.

achilleas-k reviewed (approved):
Tiny question about the variant below, but looks good.

jordigilh commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,906 @@
+	defaultImageConfig *distro.ImageConfig
+}
+
+// RHEL-based OS image configuration defaults
```
Understood

jordigilh commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,906 @@
+		name:     distro.X86_64ArchName,
+		distro:   &rd,
+		legacy:   "i386-pc",
+		bootType: distro.LegacyBootType,
```
Will update.

jordigilh commented on internal/distro/fedora/pipelines.go:
```diff
@@ -0,0 +1,670 @@
+	return p
+}
+
+func osPipeline(t *imageType,
```
I can see the line you mention about GPG keys, perhaps I added it back in one of my updates....

jordigilh replied:
I'll resolve the conversation since the line you mention is already included...
https://github.com/osbuild/osbuild-composer/blob/3b8fd155956b75d7495185d33681b8f13968ca37/internal/distro/fedora/pipelines.go#L258-L260

jordigilh commented on internal/distro/fedora/distro.go:
```diff
@@ -0,0 +1,906 @@
+			},
+		},
+		bootable:            true,
+		defaultSize:         3 * GigaByte,
```
Acknowledged. Changed to 2GB.

jordigilh commented on internal/distro/fedora/distro.go:
@thozza I can't figure out which image type belongs this configuration, I assume it's for EC2, can you confirm? I will explicitly set it to `multi-user.target`.

jordigilh replied:
Moved all image types to `multi-user.target`

jordigilh commented on internal/distro/fedora/distro.go:
Understood. I will update the value to remove the second `console` entry.

jordigilh commented:
@achilleas-k @thozza added 2 new commits based on feedback that @thozza left earlier that I forgot to address. Once the tests are green I'll ping you both again to review the 2 new commits.

ondrejbudai commented:
Manifest tests are failing, can someone take a look, please? :) I would love to see this merged.

gicmo commented:
Manifests and are GCP test:
```
Apr 23 10:11:52 ip-10-30-18-34.us-east-1.aws.redhat.com osbuild-worker[20281]: time="2022-04-23T10:11:52Z" level=info msg="[GCP] 🧹 Deleting uploaded image file: [MASKED]/image-733b6535d1732a718104754fea964193bab7a72a2d132a059290c757.tar.gz" jobId=d08c5774-3a11-4303-992f-2dbf91db601e
[3184](https://gitlab.com/redhat/services/products/image-builder/ci/osbuild-composer/-/jobs/2366562009#L3184)Apr 23 10:11:52 ip-10-30-18-34.us-east-1.aws.redhat.com osbuild-worker[20281]: time="2022-04-23T10:11:52Z" level=error msg="osbuild job failed: failed to insert image into GCE. HTTPErrorCode:400 HTTPErrorMsg:BAD REQUEST Errors:[code:\"PERMISSIONS_ERROR\"  message:\"Required 'read' permission for 'image-733b6535d1732a718104754fea964193bab7a72a2d132a059290c757.tar.gz'\"]" jobId=d08c5774-3a11-4303-992f-2dbf91db601e
```

jordigilh commented:
@ondrejbudai the manifest tests are also failing for fedora. I forgot to run the tests to regenerate them once I changed the boot type to hybrid. I'm working on it.
Also, there is a change (https://github.com/osbuild/osbuild-composer/pull/2461#discussion_r835502340) that @thozza asked me to make a while ago that I need his feedback, but he's not available until this Friday. Can someone else help here?

jordigilh commented:
> Manifest tests are failing, can someone take a look, please? :) I would love to see this merged.

I'm on it.

jordigilh commented:
Tests failing due to out capacity issues:
```
│ Error: Error requesting spot instances: InsufficientFreeAddressesInSubnet: There are not enough free addresses in subnet 'subnet-0e4e3490fbe18253b' to satisfy the requested number of instances.
```

jordigilh commented:
@achilleas-k @thozza @ondrejbudai added two new commits that include an update to the osbuild commit ID in Schutzfile for fedora and the regenerated manifests for fedora-{34,35} on x86_64 and aarch64 due to the [change](https://github.com/osbuild/osbuild-composer/commit/26a83262a8891795f47633958c0614b2482cf489) in the rpm staged struct.

achilleas-k commented:
I think this is good to go now and _I think_ this shouldn't conflict with any of the priority RHEL changes we need to merge right now, but I'm not 100% sure, so I'd like to hold off on this until https://github.com/osbuild/osbuild-composer/pull/2552 is merged.

ondrejbudai commented:
Yeah, I'm sorry, we are currently fighting some fires that have a very high priority. I think we should be able to merge this on Monday though. Thanks for understanding.

thozza reviewed (commented):
I reviewed the code and it looks good from my point of view. Thanks a lot for doing this rework. I'm not approving the PR yet, because as we discussed directly over Slack, the image test cases should be regenerated with `image-info` report to allow checking the difference in the resulting images....

jordigilh commented on internal/distro/fedora/distro.go:
Sure, I'll remove those

jordigilh commented on internal/distro/fedora/distro.go:
most probably. I'll clean up.

thozza commented on test/data/manifests/fedora_34-x86_64-fedora_iot_commit-boot.json:
```diff
@@ -11152,9 +13890,9 @@
+      "fedora-release-34-39.noarch",
       "fedora-release-common-34-39.noarch",
-      "fedora-release-identity-iot-34-39.noarch",
-      "fedora-release-iot-34-39.noarch",
```
This looks suspicious. The `fedora-release-iot` should be probably still installed...

thozza commented on test/data/manifests/fedora_34-x86_64-fedora_iot_commit_debug-boot.json:
```diff
@@ -11158,9 +13896,9 @@
+      "fedora-release-34-39.noarch",
       "fedora-release-common-34-39.noarch",
-      "fedora-release-identity-iot-34-39.noarch",
-      "fedora-release-iot-34-39.noarch",
```
ditto, this looks suspicious....

jordigilh replied:
Fixed, the packages are back

achilleas-k commented on internal/distro/fedora/distro.go:
```diff
@@ -591,6 +592,10 @@ func (t *imageType) PayloadPackageSets() []string {
 	return []string{blueprintPkgsKey}
 }
 
+func (t *imageType) PackageSetsChains() map[string][]string {
+	return t.packageSetChains
+}
+
```
Something must have gotten mixed up during the rebase because the message for commit 0408f2322e8cbded128772990cdff372215fdfe8 doesn't match the changes.

jordigilh replied:
Indeed. This is odd. I'll take a look.

jordigilh replied:
Looks like the commit I added to update the commid ID for fedora in the `Schutzfile` was impacted when I rebased and included this later on commit:
https://github.com/osbuild/osbuild-composer/commit/e3f812f6fd60386502f921fa829debf869110ec5
I can change the description of the commit since it does not apply anymore.

thozza replied:
I think that the part adding the `PackageSetsChains()` method should be actually squashed to the commit that replaced the v1 Fedora distro implementation with v2. Because without this method, the `imageType` struct won't actually implement all methods of the `ImageType` interface and the code would be basically broken in that version.

jordigilh replied:
I'll squash the commits for the changes in the code related to `PackageSetChains()`.

thozza commented on internal/distro/fedora/package_sets.go:
```diff
@@ -0,0 +1,731 @@
+}
+
+// fedora iot commit OS package set
+func iotCommitPackageSet(t *imageType) rpmmd.PackageSet {
```
this is missing https://github.com/osbuild/osbuild-composer/blob/082da070d16eb6e73c87adcdab30ef1f0d037ccd/internal/distro/fedora33/distro.go#L737

jordigilh replied:
Added back, and the package is showing now in the new manifests iot related:
```
      {
        "name": "fedora-release-common",
        "epoch": 0,
        "version": "34",
        "release": "39",
        "arch": "noarch",
        "remote_location": "https://rpmrepo.osbuild.org/v2/mirror/public/f34/f34-aarch64-updates-released-20220113/Packages/f/fedora-release-common-34-39.noarch.rpm",
        "checksum": "sha256:7f6376e3a393b2501d6c38a8b82b8ebc54265e9b4cf3bb54f3e164a9aaf078f0",
        "check_gpg": true
      },
      {
        "name": "fedora-release-identity-iot",
        "epoch": 0,
        "version": "34",
        "release": "39",
        "arch": "noarch",
        "remote_location": "https://rpmrepo.osbuild.org/v2/mirror/public/f34/f34-aarch64-updates-released-20220113/Packages/f/fedora-release-identity-iot-34-39.noarch.rpm",
        "checksum": "sha256:323cf73160530491d2a25119c9b72ee26c0f23863b1bae71f8a04af1f471656b",
        "check_gpg": true
      },
      {
        "name": "fedora-release-iot",
        "epoch": 0,
        "version": "34",
        "release": "39",
        "arch": "noarch",
        "remote_location": "https://rpmrepo.osbuild.org/v2/mirror/public/f34/f34-aarch64-updates-released-20220113/Packages/f/fedora-release-iot-34-39.noarch.rpm",
        "checksum": "sha256:c8ca801cddbe977065fe4d9bf73606f0a14532a74e4276e357698e316b7d4d73",
        "check_gpg": true
      },
```

jordigilh replied:
Updated. Please review.

achilleas-k reviewed (commented):
Everything looks good to me.  Had a run through the whole thing again though I only really focused on the newer commits.  I have one issue with a commit-diff mismatch (see below).  I think that can be squashed into the next commit.

Thanks again for all the hard work!

thozza commented on test/data/manifests/fedora_34-x86_64-oci-boot.json:
```diff
@@ -9514,15 +12161,18 @@
-      "fedora-release-cloud-34-39.noarch",
+      "fedora-release-34-39.noarch",
       "fedora-release-common-34-39.noarch",
-      "fedora-release-identity-cloud-34-39.noarch",
```
This image-info report is apparently outdated and does not correspond with the manifest. There is `fedora-release-identity-cloud-34-39.noarch.rpm` in the manifest sources, but the image-info report does not contain it.

jordigilh replied:
Will regenerate all manifests for f34 and f35 with aarch64 and x86_64. I'll pay attention on this manifest in particular.

jordigilh replied:
Fixed. Please review.

thozza reviewed (commented):
Thanks for the changes.

During the review, I noticed that some image test case manifests either still do not have the `image-info` report part or that the report is out of sync with the manifest (IOW inspecting the image build from such manifest would produce different `image-info` report).
Please make sure to regenerate image test cases for all changed distros and all architectures with `image-info`. Otherwise, it is hard to verify that the image definitions are  reasonably similar to the ones that are currently in composer.

In addition, I noticed that the current code was adding a common kernel options to all image types https://github.com/osbuild/osbuild-composer/blob/9840ef750b19b029d5a8375b57a59cd2f4a93fc6/internal/distro/fedora33/distro.go#L442-L447

I don't think that we should use the same value for all image types in your rework, however it makes sense to add some part to each image type definitions's kernel options. I propose to use values used for the same image types in `rhel-90` distro definition. What do you think @gicmo / @achilleas-k ?

achilleas-k commented:
Thanks @thozza for the in-depth review.  I missed all those details myself.

Since we need more bigger changes, I'd also like to ask for more detailed commit messages.  None of the commit messages have a description, only a subject, and sometimes the reason for the changes isn't clear from the one line summary.  It would be nice, especially on bigger commits or less obvious ones, to have a longer description in the commit messages for what's going on and why.

achilleas-k commented on internal/distro/fedora/distro.go:
Did you mean "definitely **not** correct"?

thozza replied:
> Did you mean "definitely **not** correct"?

yes 😄

jordigilh replied:
I'll revert and restore the previous one with `console=ttyS0,115200n8` as the only `console` value

jordigilh replied:
for the sake of clarity, the new value will be (once I've pushed the new commits):
```
"ro no_timer_check console=ttyS0,115200n8 biosdevname=0 net.ifnames=0"
```
this is the original value from the existing code, not from rhel90.

thozza replied:
LGTM, thanks!

jordigilh replied:
Updated.

achilleas-k commented:
> I propose to use values used for the same image types in `rhel-90` distro definition. What do you think @gicmo / @achilleas-k ?

I think that makes sense, though I'm not an expert nor have I looked into what the official images do.

thozza commented on test/data/manifests/fedora_34-x86_64-vhd-boot.json:
```diff
@@ -8833,17 +11377,17 @@
         "grub_users": "$grub_users",
-        "initrd": "/boot/initramfs-5.15.13-100.fc34.x86_64.img",
-        "linux": "/boot/vmlinuz-5.15.13-100.fc34.x86_64",
-        "options": "root=UUID=76a22bf4-f153-4541-b6c7-0332c0dfaeac ro no_timer_check net.ifnames=0 console=tty1 console=ttyS0,115200n8",
```
this one is also suspicious, but probably OK. Mainly since there was previously
```
"kernelopts": "root=UUID=76a22bf4-f153-4541-b6c7-0332c0dfaeac ro biosdevname=0 rootdelay=300 console=ttyS0 earlyprintk=ttyS0 net.ifnames=0"
```
which didn't get reflected in the `"options"`. Now it does.... 🤷

jordigilh commented:
@achilleas-k @thozza I've changed the code as suggested. I've also reworded some of the commit messages to contain more detailed insight of the changes. Let me know if you want me to resolve the comments that I've addressed or you'd rather do that yourself and use them as checkpoints for the review.

thozza commented on test/data/manifests/fedora_34-x86_64-vmdk-boot.json:
```diff
@@ -9208,17 +11839,17 @@
         "grub_users": "$grub_users",
-        "initrd": "/boot/initramfs-5.15.13-100.fc34.x86_64.img",
-        "linux": "/boot/vmlinuz-5.15.13-100.fc34.x86_64",
-        "options": "root=UUID=76a22bf4-f153-4541-b6c7-0332c0dfaeac ro no_timer_check net.ifnames=0 console=tty1 console=ttyS0,115200n8",
```
kernel options change....

jordigilh replied:
Fixed. The package is back,Updated. Please review.

thozza reviewed (commented):
Thanks for the changes.

I'm sorry, but I was probably not as precise in my wording as I should have been. While you fixed the `ami` and `qcow2`, there are still changes in the kernel options for some image types (e.g. `oci`, `vmdk`, `openstack`). Since a fixed set of kernel options was previously added to each image type, this means that each image type after the rework must have some kernel options specified in its definition.

Everything else besides the kernel options looks good and I do not see any other issues!

jordigilh commented on test/data/manifests/fedora_34-aarch64-openstack-boot.json:
```diff
@@ -9659,17 +12217,17 @@
         "grub_users": "$grub_users",
-        "initrd": "/boot/initramfs-5.15.13-100.fc34.aarch64.img",
-        "linux": "/boot/vmlinuz-5.15.13-100.fc34.aarch64",
-        "options": "root=UUID=76a22bf4-f153-4541-b6c7-0332c0dfaeac ro no_timer_check net.ifnames=0 console=tty1 console=ttyS0,115200n8",
```
Updated. Please review.

jordigilh commented on test/data/manifests/fedora_34-x86_64-qcow2-boot.json:
```diff
@@ -9207,17 +11840,17 @@
         "grub_users": "$grub_users",
-        "initrd": "/boot/initramfs-5.15.13-100.fc34.x86_64.img",
-        "linux": "/boot/vmlinuz-5.15.13-100.fc34.x86_64",
-        "options": "root=UUID=76a22bf4-f153-4541-b6c7-0332c0dfaeac ro no_timer_check net.ifnames=0 console=tty1 console=ttyS0,115200n8",
```
Updated. Please review.
